subsys/net/lib/aws_fota/src/aws_fota.c

```diff
@@ -0,0 +1,517 @@
+/* Topic for updating shadow topic with version number */
+#define UPDATE_DELTA_TOPIC AWS "%s/shadow/update"
+#define UPDATE_DELTA_TOPIC_LEN (AWS_LEN +\
+				CONFIG_CLIENT_ID_MAX_LEN +\
+				(sizeof("/shadow/update") - 1))
+#define SHADOW_STATE_UPDATE \
+	"{\"state\":{\"reported\":{\"nrfcloud__fota_v1__app_v\":\"%s\"}}}"
+
+static int update_device_shadow_version(struct mqtt_client *const client)
+{
+	struct mqtt_publish_param param;
+	char update_delta_topic[UPDATE_DELTA_TOPIC_LEN + 1];
+	u8_t shadow_update_payload[CONFIG_DEVICE_SHADOW_PAYLOAD_SIZE];
+
+	int ret = snprintf(update_delta_topic,
+			   sizeof(update_delta_topic),
+			   UPDATE_DELTA_TOPIC,
+			   client->client_id.utf8);
+	u32_t update_delta_topic_len = ret;
+
+	if (ret >= UPDATE_DELTA_TOPIC_LEN) {
+		return -ENOMEM;
+	} else if (ret < 0) {
+		return ret;
+	}
+
+	ret = snprintf(shadow_update_payload,
+		       CONFIG_DEVICE_SHADOW_PAYLOAD_SIZE,
+		       SHADOW_STATE_UPDATE,
+		       version);
+	u32_t shadow_update_payload_len = ret;
+
+	if (ret >= UPDATE_DELTA_TOPIC_LEN) {
+		return -ENOMEM;
+	} else if (ret < 0) {
+		return ret;
+	}
+
+	param.message.topic.qos = MQTT_QOS_1_AT_LEAST_ONCE;
+	param.message.topic.topic.utf8 = update_delta_topic;
+	param.message.topic.topic.size = update_delta_topic_len;
+	param.message.payload.data = shadow_update_payload;
+	param.message.payload.len = shadow_update_payload_len;
+	param.message_id = sys_rand32_get();
+	param.dup_flag = 0;
+	param.retain_flag = 0;
+
+	return mqtt_publish(client, &param);
+}
+
+#define STATUS_DETAILS_TEMPLATE "{\"nextState\":\"%s\"}"
+/*
+ * #define STATUS_DETAILS_MAX_LEN  ((sizeof("{\"nextState\":\"\"}") - 1) +\
+ *  (sizeof("download_firmware") - 1))
+ */
+#define STATUS_DETAILS_MAX_LEN 255
+
+static int update_job_execution(struct mqtt_client *const client,
+				const u8_t *job_id,
+				enum execution_status state,
+				enum fota_status next_state,
+				int version_number,
+				const char *client_token)
+{
+		char status_details[STATUS_DETAILS_MAX_LEN + 1];
+		int ret = snprintf(status_details,
+				   STATUS_DETAILS_MAX_LEN,
+				   STATUS_DETAILS_TEMPLATE,
+				   fota_status_strings[next_state]);
+		if (ret >= STATUS_DETAILS_MAX_LEN) {
+			return -ENOMEM;
+		} else if (ret < 0) {
+			return ret;
+		}
+
+		ret =  aws_jobs_update_job_execution(client,
+					      job_id,
+					      state,
+					      status_details,
+					      version_number,
+					      client_token);
+
+		if (ret < 0) {
+			LOG_ERR("aws_jobs_update_job_execution failed: %d",
+				ret);
+		}
+
+		return ret;
+}
+
+
+static int aws_fota_on_publish_evt(struct mqtt_client *const client,
+				   const u8_t *topic,
+				   u32_t topic_len,
+				   u8_t *json_payload,
+				   u32_t payload_len)
+{
+	printk("fota_published_to_sub_topic_evt handler\n");
+	int err;
+
+	/* If not processign job */
+	/* Reciving an publish on notify_next_topic could be a job */
+	if (!strncmp(notify_next_topic, topic,
+		    MIN(NOTIFY_NEXT_TOPIC_MAX_LEN, topic_len))) {
```

Comment:
Should this be changed to the following?

	if ((topic_len <= NOTIFY_NEXT_TOPIC_MAX_LEN) &&
		!strncmp(notify_next_topic, topic, topic_len))

Reply:
Hmm I can change it and add an `__ASSERT` for `topic_len > NOTIFY_NEXT_TOPCI_MAX_LEN`  as it would do an out of bounds cmp if topic_len is larger.

Reply:
That should be handled by the first line, right?

Reply:
didn't read, just saw the strncmp, but won't strncmp be checked?

Reply:
If the first statement fails, it skips executing the second.

Reply:
Okei tnx :+1: